File: skylog-core/src/iosMain/kotlin/dev/viethung/skylog/platform/OsLogWriter.kt
package dev.viethung.skylog.platform

import dev.viethung.skylog.LogEntry
import dev.viethung.skylog.LogWriter
import dev.viethung.skylog.Severity

// v0.1 uses NSLog only (simpler, public API). v0.2 may switch to os_log via cinterop
// for log-archive integration (PRD RP-06).
class OsLogWriter : LogWriter() {
    override fun log(entry: LogEntry) {
        val prefix = "[${entry.severity.name.first()}] ${entry.tag}"
        val msg = entry.message + (entry.fields?.let { " " + it.entries.joinToString(" ") { (k, v) -> "$k=$v" } } ?: "")
        platform.Foundation.NSLog("$prefix: $msg")
        entry.throwable?.let { t ->
            // Mega Plan Review §2 finding F-2.4: stackTraceToString() on K/N can throw on
            // pathological stacks; isolate so logging never fails.
            val trace = try {
                t.stackTraceToString()
            } catch (_: Throwable) {
                "<stack trace unavailable: ${t::class.simpleName}>"
            }
            platform.Foundation.NSLog("    $trace")
        }
    }
}

actual fun platformDefaultWriter(): LogWriter = OsLogWriter()

File: swift-package/SkylogKit/Sources/SkylogKit/writers/OsLogWriter.swift
import Foundation

/// A `LogWriter` that emits entries via `NSLog`.
///
/// Mirrors `dev.viethung.skylog.platform.OsLogWriter` from `:skylog-core` (iosMain).
///
/// v0.1 uses `NSLog` only — simpler public API with no privacy-redaction concern
/// (PRD R-03, RP-06). The message lambda is fully evaluated before emission so no
/// privacy-level annotation is needed. Consumers requiring `os_log`'s log-archive
/// integration or privacy redaction should subclass and override `log(_:)`.
public final class OsLogWriter: LogWriter {

    public init() {}

    public func log(_ entry: LogEntry) {
        let prefix = "[\(entry.severity.letter)] \(entry.tag)"
        var msg = entry.message
        if let fields = entry.fields, !fields.isEmpty {
            let fieldStr = fields.map { "\($0.key)=\($0.value)" }.joined(separator: " ")
            msg += " \(fieldStr)"
        }
        NSLog("%@: %@", prefix, msg)

        if let err = entry.throwable {
            NSLog("    %@: %@", String(describing: type(of: err)), err.localizedDescription)
        }
    }
}
